clang/tools/libclang/CDependencies.cpp

```diff
@@ -41,12 +41,71 @@ struct DependencyScannerServiceOptions {
 
   ScanningOutputFormat getFormat() const;
 };
+
+struct CStringsManager {
+  SmallVector<std::unique_ptr<std::vector<const char *>>> OwnedCStr;
+  SmallVector<std::unique_ptr<std::vector<std::string>>> OwnedStdStr;
+
+  template <typename StringTy>
+  CXCStringArray createCStringsRefImpl(
+      ArrayRef<StringTy> Strings,
+      std::function<const char *(const StringTy &)> strGetter) {
+    OwnedCStr.push_back(std::make_unique<std::vector<const char *>>());
+    std::vector<const char *> &CStrings = *OwnedCStr.back();
+    CStrings.reserve(Strings.size());
+    for (const auto &String : Strings)
+      CStrings.push_back(strGetter(String));
+    return {CStrings.data(), CStrings.size()};
+  }
+
+  /// The two methods below create a CXCSringArray that does not own its
+  /// contents.
+  CXCStringArray createCStringsRef(ArrayRef<std::string> Strings) {
+    return createCStringsRefImpl<std::string>(
+        Strings, [](const std::string &String) { return String.c_str(); });
+  }
+
+  CXCStringArray createCStringsRef(ArrayRef<StringRef> Strings) {
+    return createCStringsRefImpl<StringRef>(
+        Strings, [](const StringRef &String) { return String.begin(); });
```

Comment:
Nit: I think `data()` fits the semantics better than `iterator()`.

But more importantly: I think this assumes `String` is null-terminated, and bad things will happen if it's not.

Reply:
We talked offline and agreed to just make a copy of the strings here instead of chasing the non-allocating ideal.